core/internal/runwork/runwork.go

```diff
@@ -142,13 +143,32 @@ func (rw *runWork) AddWorkOrCancel(
 	// until we exit and decrement addWorkCount---so internalWork
 	// is guaranteed to not be closed until this method returns.
 
-	select {
-	case <-rw.closed:
-		// Here, Close() must have been called, so we should drop the record.
-		rw.logger.CaptureError(errRecordAfterClose, "work", work)
-
-	case <-cancel:
-	case rw.internalWork <- work:
+	start := time.Now()
+	for i := 0; ; i++ {
+		select {
+		// Detect deadlocks and hangs that prevent internalWork
+		// from flushing.
+		case <-time.After(10 * time.Minute):
+			// Stop warning after the first hour to minimize spam.
```

Comment:
do we care about recovery? if it stalled for x time and after that it was able to make progress and stalled again? or do we care that there was nothing going on for the first hour?